tasks/Task.cpp

```diff
@@ -2,48 +2,422 @@
+    configurePeerDataChannel(remote_peer_id);
+}
+
+shared_ptr<rtc::PeerConnection> Task::initiatePeerConnection(string const &remote_peer_id)
 {
+    auto peer_connection = std::make_shared<rtc::PeerConnection>(mConfig);
+
+    peer_connection->onStateChange(
+        [&](rtc::PeerConnection::State state)
+        {
+            switch (state)
+            {
+            case rtc::PeerConnection::State::Disconnected:
+                mState.peer_connection.state = Disconnected;
+            case rtc::PeerConnection::State::Closed:
+                mState.peer_connection.state = Closed;
+            case rtc::PeerConnection::State::Connected:
+                mState.peer_connection.state = Connected;
+            case rtc::PeerConnection::State::Connecting:
+                mState.peer_connection.state = Connecting;
+            case rtc::PeerConnection::State::Failed:
+                mState.peer_connection.state = Failed;
+            case rtc::PeerConnection::State::New:
+                mState.peer_connection.state = NewConnection;
+            default:
+                break;
+            }
+        });
+
+    peer_connection->onGatheringStateChange(
+        [&](rtc::PeerConnection::GatheringState state)
+        {
+            switch (state)
+            {
+            case rtc::PeerConnection::GatheringState::Complete:
+                mState.peer_connection.gathering_state = Complete;
+            case rtc::PeerConnection::GatheringState::InProgress:
+                mState.peer_connection.gathering_state = InProgress;
+            case rtc::PeerConnection::GatheringState::New:
+                mState.peer_connection.gathering_state = NewGathering;
+            default:
+                break;
+            }
+        });
+
+    peer_connection->onSignalingStateChange(
+        [&](rtc::PeerConnection::SignalingState state)
+        {
+            switch (state)
+            {
+            case rtc::PeerConnection::SignalingState::HaveLocalOffer:
+                mState.peer_connection.signaling_state = HaveLocalOffer;
+            case rtc::PeerConnection::SignalingState::HaveLocalPranswer:
+                mState.peer_connection.signaling_state = HaveLocalPranswer;
+            case rtc::PeerConnection::SignalingState::HaveRemoteOffer:
+                mState.peer_connection.signaling_state = HaveRemoteOffer;
+            case rtc::PeerConnection::SignalingState::HaveRemotePranswer:
+                mState.peer_connection.signaling_state = HaveRemotePranswer;
+            case rtc::PeerConnection::SignalingState::Stable:
+                mState.peer_connection.signaling_state = Stable;
+            default:
+                break;
+            }
+        });
+
+    peer_connection->onLocalDescription(
+        [&, remote_peer_id](rtc::Description description)
+        {
+            Json::Value message;
+            message["protocol"] = "one-to-one";
+            message["to"] = remote_peer_id;
```

Comment:
I'm a bit skeptical about this ... I think that the `remote_peer_id` is specific to webrtc, and not to our signalling protocol. Use the `from` field here too.

Generally speaking, I think you really need to create a sendSignallingMessage(message) that would fill the protocol, to and from field ... to avoid forgetting something.

Reply:
And instead of referring to mDecoder, set a mRemoteID field.

- either using the remote peer id property in case it's not empty
- or when receiving the offer message